torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,292 @@
+from collections import defaultdict
+from typing import Callable, Dict
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+from torch._subclasses.fake_tensor import disable_fake_tensor_mode_tracing
+
+aten = torch.ops.aten
+
+rng_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+offset_calculator = {}
+
+
+def throw_if_philox_offset_did_not_advance(fn):
+    # Raises error if one forgot to advance the offset while adding a new decomposition.
+    def inner(*args, **kwargs):
+        old_relative_offset = PhiloxStateTracker.get_current_relative_offset()
+        out = fn(*args, **kwargs)
+        new_relative_offset = PhiloxStateTracker.get_current_relative_offset()
+        if old_relative_offset == new_relative_offset:
+            raise ValueError("Philox seed offset did not advance after decomposition")
+        return out
+
+    return inner
+
+
+def register_rng_decomposition(aten_op):
+    def inner(fn):
+        return decomp.register_decomposition(aten_op, rng_decompositions)(
+            throw_if_philox_offset_did_not_advance(fn)
+        )
+
+    return inner
+
+
+# Offset calculations depdens on the distribution
+def register_offset_calculation(aten_ops):
+    def inner(fn):
+        nonlocal aten_ops
+        aten_ops = list(aten_ops)
+        for aten_op in aten_ops:
+            offset_calculator[aten_op] = fn
+        return fn
+
+    return inner
+
+
+def get_default_stride(size):
+    """
+    A helper function to get the strides for a contiguous tensor of a given
+    shape.
+    """
+    stride = [1] * len(size) + [1]
+    for idx in reversed(range(len(size))):
+        stride[idx] = stride[idx + 1] * size[idx]
+    stride = stride[1:]
+    return stride
+
+
+class RNGFunctionalizationError(Exception):
```

Comment:
Why is this a special error type?

Reply:
So that we can catch it in the test for negative testing.

Reply:
We usually just match the error message via regex in `assertRaisesRegex`. I would use that unless you need this type evry explicitly.